lib/elixir/lib/access.ex

```diff
@@ -31,83 +26,52 @@ defmodule Access do
-  `Access.get(term, key, nil)`.
+  The bracket syntax itself places no restrictions on the key types, but the data
+  structure being accessed may have such restrictions (ie. maps can have any key type,
+  but keyword lists must have atoms as keys).
 
-  `Access` can be combined with `Kernel.put_in/3` to put a value
-  in a given key:
+  This works well for dynamic data structures where keys can come and go over the
+  lifetime of a program _but is disabled by default for structs_, which have keys
+  defined at _compile-time_ (see: [Structs and Dot Syntax](#module-structs-and-dot-syntax)).
 
-      iex> map = %{a: 1, b: 2}
-      iex> put_in map[:a], 3
-      %{a: 3, b: 2}
+  ### Bracket Syntax and Access
 
-  This syntax is very convenient as it can be nested arbitrarily:
+  When compiling  source code, Elixir transforms each invocation of bracket syntax
+  into individual calls to `Access.get/2`.
 
-      iex> users = %{"john" => %{age: 27}, "meg" => %{age: 23}}
-      iex> put_in users["john"][:age], 28
-      %{"john" => %{age: 28}, "meg" => %{age: 23}}
+      iex> quote do a[:b] end
+      {{:., [], [Access, :get]}, [], [{:a, [], AccessTest}, :b]}
 
-  Furthermore, `Access` transparently ignores `nil` values:
+  Because `Access.get/2` is translated into `Access.get/3` with `nil` as the default,
+  accessing a non-existing key will _not raise_, but will instead return a `nil` value.
 
       iex> keywords = [a: 1, b: 2]
       iex> keywords[:c][:unknown]
       nil
 
-  Since `Access` is a behaviour, it can be implemented for key-value
-  data structures. The implementation should be added to the
-  module that defines the struct being accessed. `Access` requires the
-  key comparison to be implemented using the `===` operator.
-
-  ## Static lookups
-
-  The `Access` syntax (`data[key]`) cannot be used to access fields in
-  structs, since structs do not implement the `Access` behaviour by
-  default. It is also a design decision: the dynamic access lookup
-  is meant to be used for dynamic key-value structures, like maps
-  and keywords, and not by static ones like structs (where fields are
-  known and not dynamic).
-
-  Therefore Elixir provides a static lookup for struct fields and for atom
-  fields in maps. Imagine a struct named `User` with a `:name` field.
-  The following would raise:
+  Structs do not by default implement Access behaviour and so by default cannot use this
+  notation (see: [Structs and Dot Syntax](#module-structs-and-dot-syntax)) but
+  Elixir allows the Access behaviour to be [implemented](#module-implementing-the-behaviour)
+  for user-defined struct types after which, bracket syntax may be used.
 
-      user = %User{name: "John"}
-      user[:name]
-      # ** (UndefinedFunctionError) undefined function User.fetch/2 (User does not implement the Access behaviour)
-
-  Structs instead use the `user.name` syntax to access fields:
-
-      user.name
-      #=> "John"
+  ## Access and `Kernel`
```

Comment:
Yes, I think I see what you mean. In that sense, the docs are focused on the syntax and where you can use it.

I think one source of confusion I am trying to clear up is that dot-syntax through the general compiler is different than through `Kernel usage for put/update/pop` as the latter is supported specifically through macros. It’s one of my favorite features and led to my confusion as to why it can’t be similarly used in `Map.put()` (ie. "How come `Map.put(my_map.foo, 3)` doesn’t work?”).

One clarifying question about your example. Are all of those headings in `lib/elixir/pages/Syntax Reference.md`?

Maybe I can incorporate all this so each syntax discussion has its own headers and the detail on modules is in each module’s documentation.

---

- `lib/elixir/pages/Syntax Reference.md`
    Central Syntax discussion
    - Key access (Bracket syntax)
       - Note about compiles to Access.get/2, etc... (`link` to Access module)
       - Kernel usage for put/update/pop (`link`)
    - Field access (Dot syntax)
        - Designed for maps with atom fields
        - Discussion of structs and how it raises
       - Kernel usage for put/update/pop (`link`)
- `lib/elixir/lib/access.ex`
    Central Access discussion
    - Note that (`fetch/2`, `get_and_update/3`, `get/3`, and `pop/2`) provide the foundation for key access (`link` to key access syntax) and Kernel accessor functions (`link`)
    - Detail and examples about helper functions (`Access.all/0`) and how they can be used as keys with Kernel accessor functions (`link`)
   - Note on how to implement behaviour
- `lib/elixir/lib/kernel`
   Lots to discuss here, but these Accessor functions get their own section.
    - Accessor functions (`put_in`, etc...)
        - Can be used with anything that implements Access behaviour (`link` to Access notes on implementing behaviour)
        - Detail on that a function can be a key (`link` to Access module documentation for prebuilt ‘key’ functions)
        - Includes macros that interpret both field syntax and key syntax (`link` to syntax document)
---

Is that looking good?